crates/oro-node-semver/src/version_req.rs

```diff
@@ -0,0 +1,336 @@
+use nom::branch::alt;
+use nom::bytes::complete::tag;
+use nom::character::complete::{digit1, space0};
+use nom::combinator::{all_consuming, map, map_res, opt, recognize};
+use nom::error::{context, convert_error, ParseError, VerboseError};
+use nom::sequence::tuple;
+use nom::{Err, IResult};
+
+use crate::{Identifier, SemverError};
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+enum Operation {
+    Exact,
+    GreaterThan,
+    GreaterThanEquals,
+    LessThan,
+    LessThanEquals,
+    Compatible,
+    WildCard(WildCardVersion),
+}
+
+impl std::string::ToString for Operation {
+    fn to_string(&self) -> String {
+        use Operation::*;
+        match self {
+            GreaterThan => ">".into(),
+            GreaterThanEquals => ">=".into(),
+            LessThan => "<".into(),
+            LessThanEquals => "<=".into(),
+            _ => "***".into(),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+enum WildCardVersion {
+    Minor,
+    Patch,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub(crate) struct Predicate {
+    operation: Operation,
+    major: usize,
+    minor: Option<usize>,
+    patch: Option<usize>,
+    pre_release: Vec<Identifier>,
+}
+
+impl ToString for Predicate {
+    fn to_string(&self) -> String {
+        format!(
+            "{}{}.{}.{}",
+            self.operation.to_string(),
+            self.major,
+            self.minor.unwrap_or_else(|| 0),
+            self.patch.unwrap_or_else(|| 0),
+        )
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct VersionReq {
+    predicates: Vec<Predicate>,
+}
+
+pub fn parse<S: AsRef<str>>(input: S) -> Result<VersionReq, SemverError> {
+    let input = &input.as_ref()[..];
+
+    match all_consuming(predicates::<VerboseError<&str>>)(input) {
+        Ok((_, predicates)) => Ok(VersionReq { predicates }),
+        Err(err) => Err(SemverError::ParseError {
+            input: input.into(),
+            msg: match err {
+                Err::Error(e) => convert_error(input, e),
+                Err::Failure(e) => convert_error(input, e),
+                Err::Incomplete(_) => "More data was needed".into(),
+            },
+        }),
+    }
+}
+
+fn predicates<'a, E>(input: &'a str) -> IResult<&'a str, Vec<Predicate>, E>
+where
+    E: ParseError<&'a str>,
+{
+    context(
+        "predicate alternatives",
+        alt((hypenated_with_only_major, full_version_range)),
+    )(input)
+}
+
+fn hypenated_with_only_major<'a, E>(input: &'a str) -> IResult<&'a str, Vec<Predicate>, E>
```

Comment:
I am a little worried I am going off track here is this looks if-else heavy.

Are there better strategies to parsing this?